tiledb/sm/array/array_directory.cc

```diff
@@ -393,8 +393,37 @@ Status ArrayDirectory::load() {
           "Cannot open array; Array does not exist."));
     }
 
-    // Set the latest array schema URI
-    latest_array_schema_uri_ = array_schema_uris_.back();
+    // Set the latest array schema URI. When in READ mode, the latest array
+    // schema URI is the schema with the largest timestamp less than or equal
+    // to the current timestamp_end_. If no schema meets this definition, we
+    // use the first schema available.
+    if (mode_ == ArrayDirectoryMode::READ) {
+      optional<URI> latest_uri = nullopt;
+      uint64_t latest_ts = 0;
+
+      for (auto& uri : array_schema_uris_) {
+        auto name = uri.remove_trailing_slash().last_path_part();
+
+        // Skip the old schema URI name since it doesn't have timestamps
+        if (name == constants::array_schema_filename) {
+          continue;
+        }
+
+        std::pair<uint64_t, uint64_t> ts_range;
+        throw_if_not_ok(utils::parse::get_timestamp_range(uri, &ts_range));
+
+        if (ts_range.second > latest_ts && ts_range.second <= timestamp_end_) {
+          latest_uri = uri;
+          latest_ts = ts_range.second;
+        }
+      }
```

Comment:
Could you extract this loop to a function to somewhere in `storage_format/`? It is, after all, parsing URIs, which is something we want to centralize there regardless.

This particular one caught my eye for another reason. We're going to be adding submillisecond sequence numbers to our basic timestamps. While it won't directly impinge on this particular code, we're going to want to make everything that uses timestamps generic to avoid bleeding internals of the new abstraction into every use of it.

No need to do more than extract it. We can rewrite it later, and if it's already moved, we won't have to do so in order to proceed with the work.

Reply:
> Could you extract this loop to a function to somewhere in `storage_format/`? It is, after all, parsing URIs, which is something we want to centralize there regardless.

Not sure wha you mean here. I'm obviously calling a function to parse URIs, but I don't think that means this code is doing the parsing or belongs in `storage_format`. I had contemplated putting this logic in `ArrayDirectory::latest_array_schema_uri()` but I haven't got any clue how hot that function is. I could go back and make the `latest_array_schema_uri_` and optional and then only invoke this logic when that member is `nullopt`. However, given the logic happens exactly once I wasn't particularly concerned in just putting the logic in the one place its used given I don't see where else we'd apply it.

For the sub-millisecond sequences, I was under the impression that we were going to encode that into the UUID? Is there a design doc on the new approach?

> No need to do more than extract it. We can rewrite it later, and if it's already moved, we won't have to do so in order to proceed with the work.

Why would we need to rewrite this? Because of the sequence numbers?

Reply:
>sub-millisecond sequences, I was under the impression that we were going to encode that into the UUID

Yes, but that's not where the story ends. It's where it begins. What we're effectively doing is increasing our temporal resolution by not using just a clock. It's essentially a new temporal type that's going to end up everywhere in the code.

Reply:
> Why would we need to rewrite this? Because of the sequence numbers?

Exactly this.

The larger issue is that `storage_format` involves any of (1) a file format, (2) a file name, or (3) directory contents.